### ISSUE TITLE
Header implementation as a proper data type?

### ISSUE THREAD

**Issue by pyx** [huge-success/sanic#1246]
I have not delved into the official specification of HTTP headers (RFC etc.), please correct me if I am wrong. Thanks.

AFAIK, there are two properties that a proper header datatype differ from dictionaries:
1) case insensitive keys #1242 
2) possible multiple values for a single key

## To address 1)
there are lots of methods have to be taken care of, for example, a straight forward implementation that inherit from `dict` or adhere to the dictionary interface had to have these overridden/implemented (e.g using `key.lower()` for storage and lookup instead of `key` itself):

### Essential interface
- `__init__` for creating an instance from existing `mapping type` or from `collections of 2-tuples`
- `__getitem__` and `__setitem__` for `d[key]` access
- `__delitem__` for `del d[key]`
- `__contains__` for `key in d` and `key not in d` checking
- `get` for access with default value
- `update` for obvious reason

### Need custom implementation only if the stored internal keys are not normalized (not turning into lowercase before saving, for example)
- `keys` to list all keys
- `items` to get a view
- `iterkeys` and `iteritems` for old interface (probably not needed, as Sanic is python 3 only)

### Essential if keeping the dictionary interface
- `pop` to remove and return a value
- `setdefault` is nice to have

### Not necessary but surprising if not implemented when inherit from `dict`
- `popitem` no use for headers

## To address 2)
Either storing a list all the time, or using custom logic, depending on how the multiple values support is implemented, we might need a method to get all values for a single key, in many library's header implementation, it's usually called `getlist` or `get_all`.

## TLDR;
As you can see, there are a lots of boilerplate code if we need to get it right, so I suggested using `wsgiref.headers.Header`, or DTRT in the included implementation, `class CIDict` as it is right now, https://github.com/channelcat/sanic/blob/745a1d6e94936b593fb5d93070a6e3d7c6eca800/sanic/server.py#L42 is not good enough.

**Comment by yunstanford** [comment]
https://github.com/aio-libs/multidict is also a good choice

**Comment by ashleysommer** [comment]
I ran into this issue when porting the plugin Flask-CORS to Sanic-CORS.

Flask-CORS used inconsistent casing throughout the code, because Flask uses the werkzeug case-insensitive multidict implementation it wasn't a problem for them.

When porting the code to make the Sanic plugin, I found that Sanic's Multidict is case _sensitive_, so I created my own simple `CIDict` implementation to manage the headers for passing to a Response object.

Unfortunately a lot of uses found this broke their application. (https://github.com/ashleysommer/sanic-cors/pull/15), if they put Sanic-CORS after a different middleware in the chain. If the Response object already has a headers dict, it is in Sanic's MultiDict struct, that uses `MultiHeader` keys, rather than `str` keys. Eventually I tracked down and replaced all intances of inconsistent casing in the code, and changed it all to lower-case, then replaced my CIDict struct with a normal `dict` that _is_ compatible with sanic's MultiDict.

Long story short, for maximum compliance and usability, sanic _should_ have a light weight, high-performance, case-insensitive MultiDict built in.

**Comment by asvetlov** [comment]
IIRC Sanic used `multidict` but performance tests demonstrated 5% degradation (I didn't check it myself and have no idea how the change impacts the performance on current master).
The project prefers speed over correctness, that's maybe fine. 
aiohttp goes in opposite direction :)

**Comment by yunstanford** [comment]
@ashleysommer current `CIDict` inside `Sanic` doesn't handle all cases well, a little buggy, i've filed PR for fixing. I'm also fine to convert to `multidict`.

**Comment by vltr** [comment]
Pinging: is this already implemented in the latest version or is something still pending?

**Comment by sjsadowski** [comment]
@vltr I believe this was closed with PR#12444
